keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+anchor set into the container filesystem of my main container.  My main
+container doesn't start until `ca_certificates.pem` is present and non-empty.
+
+I ensure that my client application is capable of reading a set of trust anchors
+from a file containing PEM-wrapped DER-formatted X.509 certificates at the
+location I've specified
+(`/var/run/example-com-server-tls-trust-anchors/ca_certificates.pem`).
+
+I ensure that my client application notices changes to `ca_certificates.pem`
+(using inotify or periodic polling), and properly updates its TLS stack when
+that happens.
+
+#### Trust Anchor Configuration for Webhook Backends
+
+In my cluster, I operate a collection of validating and mutating admission
+webhooks that use a server TLS certificate issued from my company's private CA.
+Today, when my private CA introduces a new root as part of a scheduled rotation
+(or incident response), I need to track down and adjust all
+ValidatingWebhookConfiguration and MutatingWebhookConfiguration objects in my
+cluster that refer to my private CA, and update their
+`.webhooks[].clientConfig.caBundle` fields.
+
+ClusterTrustBundle objects let me centralize the management of my private CA's
+roots within my cluster.
+
+To use it, I create a new ClusterTrustBundle containing the current roots of my
+private CA.
+
+```yaml
+apiVersion: v1alpha
+kind: ClusterTrustBunddle
+metadata:
+  name: example-com-private-ca
+spec:
+  signerName: "" # This ClusterTrustBundle isn't associated with a signer, so I
+                 # leave this empty.
+status:
+  trustAnchorSets:
+  - name: default
+    pemTrustAnchors: |
+    The old root.  Bob told me it's going to be retired soon.
+    -----BEGIN CERTIFICATE-----
+    <base64>
+    -----END CERTIFICATE-----
+
+    The new root.  Alice emailed me last week that we should begin deploying it.
+    -----BEGIN CERTIFICATE-----
+    <base64>
+    -----END CERTIFICATE-----
+  conditions:
+  - type: Ready
+    status: True
+```
+
+Then, I update each ValidatingWebhookConfiguration and
+MutatingWebhookConfiguration object that currently refer to my private CA,
+removing the CABundle field and adding a reference to the ClusterTrustBundle
+object I created.
+
+```diff
+apiVersion: admissionregistration/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: my-validating-webhook
+webhooks:
+- name: frobnicator.example.com
+  clientConfig:
+    url: https://webhooks.example.com/frobnicator
+-    caBundle: "< long base64 data>"
++    trustAnchors:
++      clusterTrustBundleName: example-com-private-ca
++      trustAnchorSetName: default
+
+```
+
+Once I do so, kube-apiserver will use the root certificates it loads from the
+designated trust anchor set in order to validate the certificate presented by
+the webhook backend.
+
+When I need to perform a root rotation, I only need to modify my single
+`example-com-private-ca` ClusterTrustBundle to introduce the new root and retire
+the old root.
```

Comment:
We can maybe have a metric for this, but it would have high cardinality.

In general, it's usually sufficient to just wait a few hours for root rotations when all clients are guaranteed to have some refresh logic for their trust anchor sets.

Reply:
I've fleshed out the metrics section.  I haven't been able to come up with a low-cardinality way to do this.  The problem seems like kubelet would need to have a metric field that names the ClusterTrustBundle object, which could explode in large clusters.